### ISSUE TITLE
circular link will make getImplicitRolesForUser stackoverflow

### ISSUE THREAD

**Issue by GopherJ** [casbin/node-casbin#150]
mode.conf
```ini
[request_definition]
r = sub, obj, act

[policy_definition]
p = sub, obj, act

[role_definition]
g = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = g(r.sub, p.sub) && r.obj == p.obj && regexMatch(r.act, p.act)
```

```csv
p, book_admin , /book/1, GET
p, pen_admin , /pen/1, GET

g, *, book_admin 
g, *, pen_admin
```

patten matching func:
```
keyMatch
```

**error**: e.getImplicitRolesForUser("alice")  will cause stackoverflow

**What's happening**

**Step1**: Since we have pattern matching function and `alice` match `*`, then when we `createRole` `alice` will add `*` in her role list. While calling `e.getImplicitRolesForUser("alice")`.

The first loop, we will get alice's direct role list:

```
["*"]
```

**Step2**: Then we call `e.getImplicitRolesForUser("*")`, and in its first iteration we get the following role list:
```
["book_admin", "pen_admin"]
```

**Step3**: Then we call `e.getImplicitRolesForUser("book_admin")` and  `e.getImplicitRolesForUser("pen_admin")`, it founds `book_admin` matches `*` then it returns also `*` and the next call will return to our **step1**.

**Comment by GopherJ** [comment]
I suggest that we switch to casbin golang's solution and avoid using recursive functions even it has another problem: https://github.com/casbin/casbin/issues/446

**Comment by GopherJ** [comment]
I suggest switching to casbin golang's implementation

**Comment by nodece** [comment]
@GopherJ The have same implementation in golang. So try to do a fix in https://github.com/casbin/casbin/pull/452.

**Comment by GopherJ** [comment]
@nodece I think we should fix: https://github.com/casbin/node-casbin/blob/0e3c49c54b4bb7affa6e0e030ed342bb6cc717d2/src/enforcer.ts#L288 first, normally it cannot handle this type of link `* -> book_admin -> *`. In casbin-rs it crashed and I fixed it in: https://github.com/casbin/casbin-rs/pull/131

**Comment by nodece** [comment]
@GopherJ I submitted a patch in #155 , Could you review it?

**Comment by nodece** [comment]
fixed in #155
